### PR TITLE
Add the MergeSort

### PR DESCRIPTION
**Pull request by Sunbelife**
I have added the mergesort and its benchmark to our project, they also passed google test.

Signed-off-by: Sunbelife Sunbelife@foxmail.com

### CODE REVIEW

**Comment by review-ninja** [comment]
<a href="https://app.review.ninja/Llama-Group/Project-Llama/pull/127" target="_blank">ReviewNinja

File: llama_cpp/test/Llama_UnitTest/SortUnitTest.cpp [hunk 1]
```diff
@@ -15,17 +15,18 @@
 //  limitations under the License.
```
Comment (0xsuu) [inline llama_cpp/test/Llama_UnitTest/SortUnitTest.cpp:15]: Dont change the format before, we use 4 spaces! not 2

**Comment by 0xsuu** [comment]
ＤＯ ＮＯＴ ＣＨＡＮＧＥ ＵＮＲＥＬＡＴＥＤ ＦＩＬＥＳ！

File: llama_cpp/src/Sort/MergeSort.h [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    std::vector<SortObject<T>> tmp(high - low + 1);
+
+    while (i <= mid && j <= high) {
+      if (arr->at(i) <= arr->at(j))
+        std::swap(tmp.at(k++), arr->at(i++));
+      else
+        std::swap(tmp.at(k++), arr->at(j++));
+    }
+
+    while (i <= mid)
```
Comment (0xsuu) [inline llama_cpp/src/Sort/MergeSort.h]: Must use {

File: llama_cpp/src/Sort/MergeSort.h [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+      mergesort(arr, low, mid);
+      // right hand
+      mergesort(arr, mid + 1, high);
+      Merge(arr, low, mid, high);
+    }
+  }
+
+  void doSort(std::vector<SortObject<T>> *reformedArray) override {
+    mergesort(reformedArray, 0, reformedArray->size() - 1);
+  }
```
Comment (0xsuu) [inline llama_cpp/src/Sort/MergeSort.h:77]: Four spaces

File: llama_cpp/test/Llama_UnitTest/SortUnitTest.cpp [hunk 4]
```diff
@@ -40,289 +41,317 @@ using llama::Sort;
 //
 
 struct SortObjectTest : public ::testing::Test {
-    SortObject<int> objInt1;
-    SortObject<int> objInt2;
+  SortObject<int> objInt1;
```
Comment (0xsuu) [inline llama_cpp/test/Llama_UnitTest/SortUnitTest.cpp]: Four spaces

**Comment by 0xsuu** [comment]
We fucking use 4 spaces not 2! Still, don't change unrelated files.

File: llama_cpp/test/Llama_UnitTest/SortUnitTest.cpp [hunk 5]
```diff
@@ -208,7 +229,7 @@ TEST_F(SortCorrectnessTest, BruteForceRandomStringsCorrectness) {
     EXPECT_EQ(randomStringsCorrect, randomStrings);
 }
 
-// Testing for Insertion Sort.
+/* Testing for Insertion Sort. */
```
Comment (0xsuu) [inline llama_cpp/test/Llama_UnitTest/SortUnitTest.cpp]: Don't change this.

File: llama_cpp/test/Llama_UnitTest/SortUnitTest.cpp [hunk 6]
```diff
@@ -167,8 +183,10 @@ struct SortCorrectnessTest : public ::testing::Test {
         mContinuousData.generateContinuousData(&continuousDoubles, 10, false);
         mContinuousData.generateContinuousData(&continuousStrings, 10, false);
         mContinuousData.generateContinuousData(&continuousIntsReversed, 10, true);
-        mContinuousData.generateContinuousData(&continuousDoublesReversed, 10, true);
-        mContinuousData.generateContinuousData(&continuousStringsReversed, 10, true);
+        mContinuousData.generateContinuousData(&continuousDoublesReversed, 10,
+                                                                                     true);
+        mContinuousData.generateContinuousData(&continuousStringsReversed, 10,
+                                                                                     true);
```
Comment (0xsuu) [inline llama_cpp/test/Llama_UnitTest/SortUnitTest.cpp]: WTF are these.

File: llama_cpp/test/Llama_UnitTest/SortUnitTest.cpp [hunk 7]
```diff
@@ -159,6 +160,10 @@ struct SortCorrectnessTest : public ::testing::Test {
     llama::SelectionSort<double> doubleSe;
     llama::SelectionSort<string> stringSe;
 
+    llama::MergeSort<int> intme;
```
Comment (0xsuu) [inline llama_cpp/test/Llama_UnitTest/SortUnitTest.cpp]: intMe

File: llama_cpp/benchmark/sort_benchmark/sort_benchmark.cpp [hunk 8]
```diff
@@ -198,6 +224,15 @@ static void SeSortClass(benchmark::State& state) {
     }
 }
 
+static void MeSortClass(benchmark::State& state) {
+    llama::MergeSort<CustomisedClass> MeSort = llama::MergeSort<CustomisedClass>();
```
Comment (0xsuu) [inline llama_cpp/benchmark/sort_benchmark/sort_benchmark.cpp:228]: meSort not MeSort

File: llama_cpp/benchmark/sort_benchmark/sort_benchmark.cpp [hunk 9]
```diff
@@ -198,6 +224,15 @@ static void SeSortClass(benchmark::State& state) {
...
+    std::vector<CustomisedClass> a = {CustomisedClass(5), CustomisedClass(4),
+                          CustomisedClass(3), CustomisedClass(2), CustomisedClass(1)};
+    while (state.KeepRunning()) {
+        MeSort.performSort(&a);
```
Comment (0xsuu) [inline llama_cpp/benchmark/sort_benchmark/sort_benchmark.cpp:232]: meSort not MeSort

File: llama_cpp/src/Sort/MergeSort.h [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+  MergeSort() {}
+  explicit MergeSort(std::vector<T> *inputArray) {
+    this->performSort(inputArray);
+  }
+
+  void Merge(std::vector<SortObject<T>> *arr, int low, int mid, int high) {
+    int i, j, k;
+    i = low;
+    j = mid + 1;
+    k = 0;
```
Comment (0xsuu) [inline llama_cpp/src/Sort/MergeSort.h:39]: int i = low; not int i; i = low;
Same for j, k

File: llama_cpp/src/Sort/MergeSort.h [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+#include <algorithm>
+#include <utility>
+#include <vector>
+
+#include "Sort.h"
+
+namespace llama {
+template <class T> class MergeSort : public Sort<T> {
+public:
+  MergeSort() {}
```
Comment (0xsuu) [inline llama_cpp/src/Sort/MergeSort.h:30]: ***\* 4 ***\* spaces not 2

Comment (0xsuu) [inline llama_cpp/src/Sort/MergeSort.h:30]: Change the whole file back to 4 spaces. DO NOT USE FIND & REPLACE.

File: llama_cpp/src/Sort/MergeSort.h [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+    }
+
+    std::vector<SortObject<T>>().swap(tmp);
+  }
+
+  void mergesort(std::vector<SortObject<T>> *arr, int low, int high) {
+    int mid;
+    if (low < high) {
+      mid = (low + high) / 2;
+      // left hand
```
Comment (0xsuu) [inline llama_cpp/src/Sort/MergeSort.h]: Left hand side.

File: llama_cpp/src/Sort/MergeSort.h [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+      mergesort(arr, low, mid);
+      // right hand
```
Comment (0xsuu) [inline llama_cpp/src/Sort/MergeSort.h]: Right hand side.

File: llama_cpp/test/Llama_UnitTest/SortUnitTest.cpp [hunk 14]
```diff
@@ -192,7 +197,7 @@ struct SortCorrectnessTest : public ::testing::Test {
     }
 };
 
-// Testing for Brute Force Sort.
+/* Testing for Brute Force Sort. */
```
Comment (0xsuu) [inline llama_cpp/test/Llama_UnitTest/SortUnitTest.cpp]: Use //

File: llama_cpp/test/Llama_UnitTest/SortUnitTest.cpp (hunk 5 above)
Comment (0xsuu) [reply llama_cpp/test/Llama_UnitTest/SortUnitTest.cpp]: Use //

File: llama_cpp/test/Llama_UnitTest/SortUnitTest.cpp [hunk 15]
```diff
@@ -224,7 +229,7 @@ TEST_F(SortCorrectnessTest, InsertionRandomStringsCorrectness) {
     EXPECT_EQ(randomStringsCorrect, randomStrings);
 }
 
-// Testing for Bubble Sort.
+/* Testing for Bubble Sort. */
```
Comment (0xsuu) [inline llama_cpp/test/Llama_UnitTest/SortUnitTest.cpp]: Use //

File: llama_cpp/test/Llama_UnitTest/SortUnitTest.cpp [hunk 16]
```diff
@@ -240,7 +245,7 @@ TEST_F(SortCorrectnessTest, BubbleRandomStringsCorrectness) {
     EXPECT_EQ(randomStringsCorrect, randomStrings);
 }
 
-// Testing for Selection Sort.
+/* Testing for Selection Sort. */
```
Comment (0xsuu) [inline llama_cpp/test/Llama_UnitTest/SortUnitTest.cpp]: Use //

File: llama_cpp/test/Llama_UnitTest/SortUnitTest.cpp [hunk 17]
```diff
@@ -256,6 +261,21 @@ TEST_F(SortCorrectnessTest, SelectionRandomStringsCorrectness) {
     EXPECT_EQ(randomStringsCorrect, randomStrings);
 }
 
+/* Testing for Merge Sort. */
```
Comment (0xsuu) [inline llama_cpp/test/Llama_UnitTest/SortUnitTest.cpp]: Use //

**Pull request closed without merging**